File: csharp/Link.Foundation.Links.Notation.Tests/MultiRefTests.cs
using System;
using System.Linq;
using Xunit;

namespace Link.Foundation.Links.Notation.Tests
{
    /// <summary>
    /// Multi-Reference Feature Tests (Issue #184)
    /// Tests for multi-word references without quotes:
    /// - (some example: some example is a link)
    /// - Ids as array: ["some", "example"]
    /// - Id property throws for multi-refs, use Ids instead
    /// </summary>
    public static class MultiRefTests
    {
        [Fact]
        public static void ParsesTwoWordMultiReferenceId()
        {
            var parser = new Parser();
            var result = parser.Parse("(some example: value)");
            Assert.Single(result);
            // Use Ids property for multi-references
            Assert.NotNull(result[0].Ids);
            Assert.Equal(2, result[0].Ids.Count);
            Assert.Equal(new[] { "some", "example" }, result[0].Ids);
            Assert.Single(result[0].Values);
        }

        [Fact]
        public static void ParsesThreeWordMultiReferenceId()
        {
            var parser = new Parser();
            var result = parser.Parse("(new york city: value)");
            Assert.Single(result);
            Assert.NotNull(result[0].Ids);
            Assert.Equal(3, result[0].Ids.Count);
            Assert.Equal(new[] { "new", "york", "city" }, result[0].Ids);
        }

        [Fact]
        public static void IdPropertyThrowsForMultiRef()
        {
            var parser = new Parser();
            var result = parser.Parse("(some example: value)");
            Assert.Single(result);
            // Id property should throw MultiReferenceException for multi-refs
            var ex = Assert.Throws<MultiReferenceException>(() => result[0].Id);
            Assert.Equal(2, ex.ReferenceCount);
            Assert.Contains("Use the 'Ids' property instead of 'Id'", ex.Message);
        }

        [Fact]
        public static void ParsesSingleWordIdBackwardCompatible()
        {
            var parser = new Parser();
            var result = parser.Parse("(papa: value)");
            Assert.Single(result);
            // Single-word: Id returns string, Ids returns array with single element
            Assert.Equal("papa", result[0].Id);
            Assert.NotNull(result[0].Ids);
            Assert.Single(result[0].Ids);
            Assert.Equal(new[] { "papa" }, result[0].Ids);
        }

        [Fact]
        public static void ParsesQuotedMultiWordIdBackwardCompatible()
        {
            var parser = new Parser();
            var result = parser.Parse("('some example': value)");
            Assert.Single(result);
            // Quoted multi-word is a single reference, so Id works
            Assert.Equal("some example", result[0].Id);
            Assert.NotNull(result[0].Ids);
            Assert.Single(result[0].Ids);
            Assert.Equal(new[] { "some example" }, result[0].Ids);
        }

        [Fact]
        public static void FormatMultiReferenceId()
        {
            var parser = new Parser();
            var result = parser.Parse("(some example: value)");
            var formatted = result.Format();
            // Multi-reference IDs are formatted with quotes (normalized form)
            Assert.Equal("('some example': value)", formatted);
        }

        [Fact]
        public static void RoundTripMultiReference()
        {
            var parser = new Parser();
            var input = "(new york city: great)";
            var result = parser.Parse(input);
            var formatted = result.Format();
            // Round-trip normalizes multi-word ID to quoted form
            Assert.Equal("('new york city': great)", formatted);
        }

        [Fact]
        public static void ParsesIndentedSyntaxMultiReference()
        {
            var parser = new Parser();
            var input = "some example:\n  value1\n  value2";
            var result = parser.Parse(input);
            Assert.Single(result);
            Assert.NotNull(result[0].Ids);
            Assert.Equal(new[] { "some", "example" }, result[0].Ids);
            Assert.Equal(2, result[0].Values?.Count);
        }

        [Fact]
        public static void BackwardCompatibilitySingleLine()
        {
            var parser = new Parser();
            var result = parser.Parse("papa: loves mama");
            Assert.Single(result);
            Assert.Equal("papa", result[0].Id);
            Assert.Equal(2, result[0].Values?.Count);
        }

        [Fact]
        public static void BackwardCompatibilityParenthesized()
        {
            var parser = new Parser();
            var result = parser.Parse("(papa: loves mama)");
            Assert.Single(result);
            Assert.Equal("papa", result[0].Id);
            Assert.Equal(2, result[0].Values?.Count);
        }

        [Fact]
        public static void BackwardCompatibilityNested()
        {
            var parser = new Parser();
            var result = parser.Parse("(outer: (inner: value))");
            Assert.Single(result);
            Assert.Equal("outer", result[0].Id);
            Assert.Single(result[0].Values);
            Assert.Equal("inner", result[0].Values?[0].Id);
        }

        [Fact]
        public static void MultiRefWithMultipleValues()
        {
            var parser = new Parser();
            var result = parser.Parse("(some example: one two three)");
            Assert.Single(result);
            // Ids should contain the multi-ref parts
            Assert.NotNull(result[0].Ids);
            Assert.Equal(new[] { "some", "example" }, result[0].Ids);
            // Id should throw for multi-ref
            Assert.Throws<MultiReferenceException>(() => result[0].Id);
            // Values should be 3 separate references
            Assert.Equal(3, result[0].Values?.Count);
            Assert.Equal("one", result[0].Values?[0].Id);
            Assert.Equal("two", result[0].Values?[1].Id);
            Assert.Equal("three", result[0].Values?[2].Id);
        }

        [Fact]
        public static void MultiRefValuesAreSeparateReferences()
        {
            // Per issue #184 feedback: context-aware parsing is out of scope
            var parser = new Parser();
            var result = parser.Parse("(some example: some example is a link)");
            Assert.NotNull(result[0].Ids);
            Assert.Equal(new[] { "some", "example" }, result[0].Ids);
            // Values should be 5 separate references (no context-aware grouping)
            Assert.Equal(5, result[0].Values?.Count);
            Assert.Equal("some", result[0].Values?[0].Id);
            Assert.Equal("example", result[0].Values?[1].Id);
            Assert.Equal("is", result[0].Values?[2].Id);
            Assert.Equal("a", result[0].Values?[3].Id);
            Assert.Equal("link", result[0].Values?[4].Id);
        }
    }
}

File: rust/tests/multi_ref_tests.rs
//! Multi-Reference Feature Tests (Issue #184)
//!
//! Tests for multi-word references without quotes:
//! - (some example: some example is a link)
//! - ID as multi-word string: "some example"

use links_notation::{format_links, parse_lino_to_links, LiNo};

#[test]
fn test_parses_two_word_multi_reference_id() {
    let result = parse_lino_to_links("(some example: value)").expect("Failed to parse");
    assert_eq!(result.len(), 1);
    match &result[0] {
        LiNo::Link { ids, values } => {
            // Multi-word ID is now stored as separate words
            assert_eq!(ids.as_ref().unwrap(), &vec!["some".to_string(), "example".to_string()]);
            assert_eq!(values.len(), 1);
        }
        _ => panic!("Expected Link"),
    }
}

#[test]
fn test_parses_three_word_multi_reference_id() {
    let result = parse_lino_to_links("(new york city: value)").expect("Failed to parse");
    assert_eq!(result.len(), 1);
    match &result[0] {
        LiNo::Link { ids, .. } => {
            assert_eq!(ids.as_ref().unwrap(), &vec!["new".to_string(), "york".to_string(), "city".to_string()]);
        }
        _ => panic!("Expected Link"),
    }
}

#[test]
fn test_single_word_id_backward_compatible() {
    let result = parse_lino_to_links("(papa: value)").expect("Failed to parse");
    assert_eq!(result.len(), 1);
    match &result[0] {
        LiNo::Link { ids, .. } => {
            assert_eq!(ids.as_ref().unwrap(), &vec!["papa".to_string()]);
        }
        _ => panic!("Expected Link"),
    }
}

#[test]
fn test_quoted_multi_word_id_backward_compatible() {
    let result = parse_lino_to_links("('some example': value)").expect("Failed to parse");
    assert_eq!(result.len(), 1);
    match &result[0] {
        LiNo::Link { ids, .. } => {
            // Quoted ID should be preserved as-is
            assert_eq!(ids.as_ref().unwrap(), &vec!["some example".to_string()]);
        }
        _ => panic!("Expected Link"),
    }
}

#[test]
fn test_format_multi_reference_id() {
    let result = parse_lino_to_links("(some example: value)").expect("Failed to parse");
    let formatted = format_links(&result);
    // Multi-reference IDs are formatted with space-separated words (new behavior)
    // The formatted output may keep them unquoted if the formatter supports it
    assert_eq!(formatted, "(some example: value)");
}

#[test]
fn test_round_trip_multi_reference() {
    let input = "(new york city: great)";
    let result = parse_lino_to_links(input).expect("Failed to parse");
    let formatted = format_links(&result);
    // Round-trip preserves the multi-word ID structure
    assert_eq!(formatted, "(new york city: great)");
}

#[test]
fn test_indented_syntax_multi_reference() {
    let input = "some example:\n  value1\n  value2";
    let result = parse_lino_to_links(input).expect("Failed to parse");
    assert_eq!(result.len(), 1);
    match &result[0] {
        LiNo::Link { ids, values } => {
            assert_eq!(ids.as_ref().unwrap(), &vec!["some".to_string(), "example".to_string()]);
            assert_eq!(values.len(), 2);
        }
        _ => panic!("Expected Link"),
    }
}

#[test]
fn test_values_include_multi_reference_context() {
    // When the same multi-word pattern appears in values,
    // it should be formatted consistently
    let input = "(some example: some example is a link)";
    let result = parse_lino_to_links(input).expect("Failed to parse");
    assert_eq!(result.len(), 1);
    match &result[0] {
        LiNo::Link { ids, values } => {
            assert_eq!(ids.as_ref().unwrap(), &vec!["some".to_string(), "example".to_string()]);
            // Values should include "some", "example", "is", "a", "link"
            // (context-aware grouping not implemented in Rust yet)
            assert!(values.len() >= 4);
        }
        _ => panic!("Expected Link"),
    }
}

#[test]
fn test_backward_compatibility_single_line() {
    let result = parse_lino_to_links("papa: loves mama").expect("Failed to parse");
    assert_eq!(result.len(), 1);
    match &result[0] {
        LiNo::Link { ids, values } => {
            assert_eq!(ids.as_ref().unwrap(), &vec!["papa".to_string()]);
            assert_eq!(values.len(), 2);
        }
        _ => panic!("Expected Link"),
    }
}

#[test]
fn test_backward_compatibility_parenthesized() {
    let result = parse_lino_to_links("(papa: loves mama)").expect("Failed to parse");
    assert_eq!(result.len(), 1);
    match &result[0] {
        LiNo::Link { ids, values } => {
            assert_eq!(ids.as_ref().unwrap(), &vec!["papa".to_string()]);
            assert_eq!(values.len(), 2);
        }
        _ => panic!("Expected Link"),
    }
}

#[test]
fn test_backward_compatibility_nested() {
    let result = parse_lino_to_links("(outer: (inner: value))").expect("Failed to parse");
    assert_eq!(result.len(), 1);
    match &result[0] {
        LiNo::Link { ids, values } => {
            assert_eq!(ids.as_ref().unwrap(), &vec!["outer".to_string()]);
            assert_eq!(values.len(), 1);
            match &values[0] {
                LiNo::Link {
                    ids: inner_ids,
                    values: inner_values,
                } => {
                    assert_eq!(inner_ids.as_ref().unwrap(), &vec!["inner".to_string()]);
                    assert_eq!(inner_values.len(), 1);
                }
                _ => panic!("Expected nested Link"),
            }
        }
        _ => panic!("Expected Link"),
    }
}
